src/uct/ib/efa/ib_efa_md.c

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <uct/ib/efa/ib_efa.h>
+#include <uct/ib/base/ib_md.h>
+#include <ucs/type/status.h>
+#include <ucs/debug/log.h>
+
+
+static uint64_t
+uct_ib_efadv_access_flags(const uct_ib_efadv_t *efadv)
+{
+    uint64_t access_flags = IBV_ACCESS_LOCAL_WRITE;
+
+#ifdef HAVE_DECL_EFA_DV_RDMA_READ
+    if (efadv->efadv_attr.device_caps & EFADV_DEVICE_ATTR_CAPS_RDMA_READ) {
+        access_flags |= IBV_ACCESS_REMOTE_READ;
+    }
+#endif
+
+    return access_flags;
+}
+
+static uct_ib_md_ops_t uct_ib_efa_md_ops;
+
+static ucs_status_t uct_ib_efa_md_open(struct ibv_device *ibv_device,
+                                       const uct_ib_md_config_t *md_config,
+                                       uct_ib_md_t **p_md)
+{
+    ucs_status_t status;
+    uct_ib_device_t *dev;
+    uct_ib_efadv_md_t *md;
+
+    md = ucs_calloc(1, sizeof(*md), "ib_efa_md");
+    if (md == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    /* Open verbs context */
+    dev              = &md->super.dev;
+    dev->ibv_context = ibv_open_device(ibv_device);
+    if (dev->ibv_context == NULL) {
+        ucs_diag("ibv_open_device(%s) failed: %m",
+                 ibv_get_device_name(ibv_device));
+        status = UCS_ERR_IO_ERROR;
+        goto err;
+    }
```

Comment:
i think we can move ibv_open_device into uct_ib_verbs_md_open_common

Reply:
I first moved it there. But, having it here allows us to do the check for EFA device sooner, because uct_ib_efadv_query() requires an ibv_context. So, if I move it to uct_ib_verbs_md_open_common, then for every non-EFA device, we will run all code in uct_ib_verbs_md_open_common before realizing that the device is not EFA.

Reply:
Moved it to uct_ib_verbs_md_open_common and added a uct_ib_efadv_check herlper function to check for EFA device.